### PR TITLE
psyad mixed precision kernels (closes 1557)

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for first review from @arporter

File: src/psyclone/psyad/tl2ad.py [hunk 1]
```diff
@@ -262,29 +262,56 @@ def generate_adjoint(tl_psyir, active_variables):
...
+                adj_kernel_name, symbol_type=RoutineSymbol,
+                visibility=kernel_sym.visibility)
+            container.symbol_table.remove(kernel_sym)
+            routine.name = adj_kernel_sym.name
+        else:
+            routine.name = routine.symbol_table.next_available_name(
+                _create_adjoint_name(routine.name))
 
-    logger.debug("AD kernel will be named '%s'", routine.name)
+            logger.debug("AD kernel will be named '%s'", routine.name)
```
Comment (arporter) [inline src/psyclone/psyad/tl2ad.py:314]: Should this be un-indented one level - at the moment it's under the `else` clause?

File: src/psyclone/psyad/tl2ad.py [hunk 2]
```diff
@@ -262,29 +262,56 @@ def generate_adjoint(tl_psyir, active_variables):
...
-        # A symbol's name is immutable so create a new RoutineSymbol
-        adj_kernel_sym = container.symbol_table.new_symbol(
-            adj_kernel_name, symbol_type=RoutineSymbol,
-            visibility=kernel_sym.visibility)
-        container.symbol_table.remove(kernel_sym)
-        routine.name = adj_kernel_sym.name
-    else:
-        routine.name = routine.symbol_table.next_available_name(
-            _create_adjoint_name(routine.name))
+    # TODO issue #1820 support separate LFRic-specific and generic
```
Comment (arporter) [inline src/psyclone/psyad/tl2ad.py:265]: This is being done in #1782.

File: src/psyclone/tests/psyad/tl2ad_test.py [hunk 3]
```diff
@@ -355,6 +355,60 @@ def test_generate_adjoint_kind(fortran_reader):
...
+        "    psyir_tmp = 0.0\n"
+        "    psyir_tmp_1 = 0.0\n"
+        "    psyir_tmp_1 = psyir_tmp_1 + psyir_tmp\n"
+        "    psyir_tmp = 0.0\n\n"
+        "  end subroutine adj_kern3\n\n"
+        "end module test_mod\n")
+    psyir = FileContainer("test_file")
+    kern_list = []
+    for name in ["kern1", "kern2", "kern3"]:
+        routine_symbol_table = SymbolTable()
```
Comment (arporter) [inline src/psyclone/tests/psyad/tl2ad_test.py:396]: It's a minor point but it seems odd to construct the input code like this but have the expected output specified in monolithic fashion. Could you construct the Fortran for each kernel in this loop too?

**Review by arporter — CHANGES_REQUESTED** [review]
Just a couple of minor things.
All tests and examples are OK with compilation.
Changes are pylint clean.
The documentation probably needs to be updated to describe what happens when there are multiple TL routines.

File: src/psyclone/psyad/tl2ad.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/psyad/tl2ad.py:314]: Yes, well noticed.

File: src/psyclone/psyad/tl2ad.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/psyad/tl2ad.py:265]: Referenced this issue.

Comment (rupertford) [reply src/psyclone/psyad/tl2ad.py:265]: And closed #1820.

File: src/psyclone/tests/psyad/tl2ad_test.py (hunk 3 above)
Comment (rupertford) [reply src/psyclone/tests/psyad/tl2ad_test.py:396]: I've made it all monolithic as it was simpler. I hope that is OK.

Comment (rupertford) [reply src/psyclone/tests/psyad/tl2ad_test.py:396]: And it tests the full module rather than testing each routine at a time.

**Comment by rupertford** [comment]
Ready for next review from @arporter

File: src/psyclone/tests/psyad/tl2ad_test.py [hunk 4]
```diff
@@ -389,23 +405,11 @@ def test_generate_adjoint_multi_kernel(fortran_writer):
...
-        ad_routine = generate_adjoint(routine, [symbol.name, symbol2.name])
-        kern_list.append(ad_routine)
-    container = Container.create("test_mod", SymbolTable(), kern_list)
-    psyir.children.append(container)
-    ad_fortran_str = fortran_writer(psyir)
+        "end module adj_test_mod\n")
+    psyir = fortran_reader.psyir_from_source(tl_fortran_str)
+    ad_psyir = generate_adjoint(psyir, ["psyir_tmp", "psyir_tmp_1"])
+    ad_fortran_str = fortran_writer(ad_psyir)
+    print(ad_fortran_str)
```
Comment (arporter) [inline src/psyclone/tests/psyad/tl2ad_test.py:412]: This line needs removing.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Tests and examples are OK with compilation.
PSyAD UG has been updated and looks good.
There's one minor thing to fix (a `print` in the test) but I'll do that as part of the final merge.

File: src/psyclone/tests/psyad/tl2ad_test.py (hunk 3 above)
Comment (arporter) [reply src/psyclone/tests/psyad/tl2ad_test.py]: OK, thanks.

File: src/psyclone/tests/psyad/tl2ad_test.py (hunk 4 above)
Comment (arporter) [reply src/psyclone/tests/psyad/tl2ad_test.py:412]: and I've done so.

**Pull request merged**